### PR TITLE
ADD: File upload option for additional feature sel.

### PR DESCRIPTION
**Pull request by furkanmtorun**
Hi Max,

This is one of the features I was looking for a while #4 .
The point is to make it easy to select several additional features by uploading file (like in excluded features)
Screenshot 2023-03-27 at 22 48 21

### CODE REVIEW

File: omiclearn/utils/ui_components.py [hunk 1]
```diff
@@ -608,45 +608,79 @@ def _generate_eda_section(state):
...
+            state["additional_features"] = multiselect(
+                "Select additional features for training:",
+                help="Select additional features for training:",
+                options=state.remainder,
+                default=additional_features_df_list,
+            )
+        else:
+            state["additional_features"] = multiselect(
+                "Select additional features for training:",
+                help="Select additional features to be includedfor training:",
```
Comment (straussmaximilian) [inline omiclearn/utils/ui_components.py]: missing space here

File: omiclearn/utils/ui_components.py [hunk 2]
```diff
@@ -608,45 +608,79 @@ def _generate_eda_section(state):
...
+                "Additional features to be included for training"
+            ]
+            st.table(additional_features_df)
+            additional_features_df_list = list(
+                additional_features_df.iloc[:, 0].unique()
+            )
+            state["additional_features"] = multiselect(
+                "Select additional features for training:",
+                help="Select additional features for training:",
+                options=state.remainder,
```
Comment (straussmaximilian) [inline omiclearn/utils/ui_components.py:646]: Does this work if someone uploads entries that are not in the remainder list?

**Review by straussmaximilian — COMMENTED** [review]
Codechanges look good.

For this feature: How do you see the tradeoff for an additional UI element for this additional functionality? The motivation behind exlcude_features is that we have thousands of proteins and likely a lot to filter, but for the additional features, we first need to manually add them to an output file in the first place, so could this lead to a confusing user flow?

File: omiclearn/utils/ui_components.py (hunk 1 above)
Comment (furkanmtorun) [reply omiclearn/utils/ui_components.py]: Fixed.

File: omiclearn/utils/ui_components.py (hunk 2 above)
Comment (furkanmtorun) [reply omiclearn/utils/ui_components.py:646]: Really nice point. Thanks a lot Max!

It was throwing error (we also missed this in `Exclude features`)
Now, I fixed both sides.

Screenshot 2023-03-28 at 21 59 53
Screenshot 2023-03-28 at 22 00 41
Screenshot 2023-03-28 at 22 10 12

**Comment by furkanmtorun** [comment]
Hi @straussmaximilian ,

Thanks for your great points! 

- Now, I fixed the bug introduced with features containing file uploads (see the comment above with SS) -- that was also causing error in Exclude features but now all are resolved
- Regarding file uploading for additional features, the motivation is the following:
    - Assume that you have 30 additional (clinical) features such as `_age` , `_sex`, `_diabetes`, `_treated_with_drug_name_A`, `_treated_with_drug_name_A`, `_treated_with_drug_name_C` etc. and
       -  you would like to upload all of them at once OR
       - you would like to try different combinations such as using 15 of them firstly and then try the other half.
    - So, this is a lot of manual work to add each 30 or 15 additional features one-by-one. 
     I hope this makes it clear :)

**Comment by straussmaximilian** [comment]
Hi, yes I understand the motivation, however, what I was trying to get at is that I haven't seen so many additional features in practice yet. The maximum that I have seen so far was 9.  Was wondering if you had any recent users that wanted this feature or so. In any case, this was just for consideration, and feel free to merge the PR.

**Comment by furkanmtorun** [comment]
Thanks for your review @straussmaximilian !

> I haven't seen so many additional features in practice yet. 

I think it also depends on how you define "additional features". Biochemical features could be treated as `additional feature` (to separate from proteins or mRNAs) and plus with clinical features. So, the number can go higher.
I still think it is a valuable feature.

Cheers!

**Pull request merged**